src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                    return false;
+                }
+
+                var typeArguments = typeSymbol.GetTypeArguments();
+                if (typeArguments.Length != 2)
+                {
+                    return false;
+                }
+
+                return typeArguments.Last()?.SpecialType == SpecialType.System_Boolean &&
+                    SignatureComparer.Instance.HaveSameSignature(typeArguments.First(), typeArgument, caseSensitive: true);
+            }
+
+            private bool TryProcessFromClause(FromClauseSyntax fromClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(fromClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol == null || // First FromClause may have no OperationInfo symbols.
+                    (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol && methodSymbol.Parameters.Length == 2))
+                {
+                    processingState.AddVariableName(fromClause.Identifier.ValueText);
+                    processingState.AddStatementProcessingMethod(
+                        s => SyntaxFactory.ForEachStatement(
+                                    fromClause.Type ?? VarNameIdentifier,
+                                    fromClause.Identifier,
+                                    fromClause.Expression.WithoutTrivia(), WrapWithBlock(s)));
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessLetClause(LetClauseSyntax letClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(letClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 1)
+                {
+                    processingState.AddVariableName(letClause.Identifier.ValueText);
+                    processingState.AddStatementProcessingMethod(
+                        s => AddToBlockTop(
+                                SyntaxFactory.LocalDeclarationStatement(
+                                    SyntaxFactory.VariableDeclaration(
+                                        VarNameIdentifier,
+                                        SyntaxFactory.SingletonSeparatedList(
+                                            SyntaxFactory.VariableDeclarator(
+                                                letClause.Identifier,
+                                                argumentList: null,
+                                                SyntaxFactory.EqualsValueClause(letClause.Expression))))),
+                                s));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessJoinClause(JoinClauseSyntax joinClause, ProcessingState processingState)
```

Comment:
>TryProcessJoinClause [](start = 25, length = 20)

It would be good to add a comment describing what kind of transformation we are trying to do here